securedrop_client/storage.py

```diff
@@ -741,6 +813,54 @@ def delete_single_submission_or_reply_on_disk(
             pass
 
 
+def delete_local_conversation_by_source_uuid(session: Session, uuid: str, data_dir: str) -> None:
+    """
+    Delete local conversation belonging to source with a given UUID.
+    Add source UUID to table that tracks local db changes. Do not delete the source.
+    """
+    source = session.query(Source).filter_by(uuid=uuid).one_or_none()
+    if source:
+
+        # Find all associated replies and messages, and delete them
+        logger.debug("Delete records for source {} from local database.".format(uuid))
+        _delete_local_source_records(session, source)
+
+        # Delete all source files on disk
+        logger.debug("Delete files on disk for source {} from local database.".format(uuid))
+        delete_source_collection(source.journalist_filename, data_dir)
+
+        # Flag the conversation as having been deleted locally
+        flagged_conversation = DeletedConversation(uuid=uuid)
+        logger.debug("Add source {} to deleted conversation table (avoid re-downloading conversation)".format(uuid))
+        session.add(flagged_conversation)
+
+        session.commit()
+
+
+def _delete_local_source_records(session: Session, source: Source) -> None:
+    """
+    Delete messages, replies, and files records belonging to a given source.
+    """
+    try:
+        logger.debug("Locally delete messages associated with source {}".format(source.uuid))
+        session.query(Message).filter_by(source_id=source.id).delete()
+
+        logger.debug("Locally delete replies associated with source {}".format(source.uuid))
+        session.query(Reply).filter_by(source_id=source.id).delete()
+
+        logger.debug("Locally delete drafts associated with source {}".format(source.uuid))
+        session.query(DraftReply).filter_by(source_id=source.id).delete()
+
+        logger.debug("Locally delete file records associated with source {}".format(source.uuid))
+        session.query(File).filter_by(source_id=source.id).delete()
+
```

Comment:
The `DeletedConversation` record should be added here to be part of the same commit to ensure that a stale sync will always skip this source that has been deleted locally.